src/NHibernate.Test/Criteria/Lambda/ExpressionProcessorFindValueFixture.cs

```diff
@@ -0,0 +1,292 @@
+			var actual = GetValue(expression);
+
+			//Check with expression compile and invoke
+			var expected = expression.Compile().Invoke();
+
+			Assert.AreEqual(expected, actual);
+		}
+
+		[Test]
+		public void NullToIntegerCastFails()
+		{
+			object value = null;
+			Expression<Func<int>> expression = () => (int) value;
+
+			Assert.Throws<NullReferenceException>(() => GetValue(expression));
+
+			//Check with expression compile and invoke
+			Assert.Throws<NullReferenceException>(() => expression.Compile().Invoke());
+		}
+
+		[Test]
+		public void NullableIntegerToIntegerCastFails()
+		{
+			int? value = null;
+			Expression<Func<int>> expression = () => (int) value;
+
+			Assert.Throws<InvalidOperationException>(() => GetValue(expression));
+
+			//Check with expression compile and invoke
+			Assert.Throws<InvalidOperationException>(() => expression.Compile().Invoke());
+		}
+
+		[Test]
+		public void NullableIntegerToIntegerCast()
+		{
+			int? value = 1;
+			Expression<Func<int>> expression = () => (int) value;
+
+			var actual = GetValue(expression);
+
+			//Check with expression compile and invoke
+			var expected = expression.Compile().Invoke();
+
+			Assert.AreEqual(expected, actual);
+		}
+
+		[Test]
+		public void NullableIntegerToNullableLongImplicitCast()
+		{
+			int? value = 1;
+			Expression<Func<long?>> expression = () => value;
+
+			var actual = GetValue(expression);
+
+			//Check with expression compile and invoke
+			var expected = expression.Compile().Invoke();
+
+			Assert.AreEqual(expected, actual);
+		}
+
+		[Test]
+		public void IntegerToIntegerCast()
+		{
+			int value = 1;
+			Expression<Func<int>> expression = () => (int) value;
+
+			var actual = GetValue(expression);
+
+			//Check with expression compile and invoke
+			var expected = expression.Compile().Invoke();
+
+			Assert.AreEqual(expected, actual);
+		}
+
+		[Test]
+		public void IntegerToNullableIntegerImplicitCast()
+		{
+			int value = 12345;
+			Expression<Func<int?>> expression = () => value;
+
+			var actual = GetValue(expression);
+
+			//Check with expression compile and invoke
+			var expected = expression.Compile().Invoke();
+
+			Assert.AreEqual(expected, actual);
+		}
+
+		[Test]
+		public void IntegerToNullableLongImplicitCast()
+		{
+			int value = 12345;
+			Expression<Func<long?>> expression = () => value;
+
+			var actual = GetValue(expression);
+
+			//Check with expression compile and invoke
+			var expected = expression.Compile().Invoke();
+
+			Assert.AreEqual(expected, actual);
+		}
+
+		[Test]
+		public void IntegerToNullableDecimalImplicitCast()
+		{
+			int value = 12345;
+			Expression<Func<decimal?>> expression = () => value;
+
+			var actual = GetValue(expression);
+
+			//Check with expression compile and invoke
+			var expected = expression.Compile().Invoke();
+
+			Assert.AreEqual(expected, actual);
+		}
+
+		[Test]
+		public void Int16ToIntegerImplicitCast()
```

Comment:
This would fall into 'lambda compile' path. Do we want to fix it?

Reply:
I have tried to fix it bu convert is to much tricky dou you have any idea to fix it ?

Reply:
As I said in https://github.com/nhibernate/nhibernate-core/pull/2957#discussion_r764676198

Let's start from something simple but safe and make it more complex and sophisticated in other PRs. Unless you already have in mind some simple fix that you want to stick in this PR.

Reply:
> It is a variation of ConvertType

I would do it in another PR. It doesn't look as straightforward fix to me. For instance Boolean type and value overflows might behave differently in runtime vs ChangeType.

Reply:
I agree with @bahusoid, in my test I was stuck with "NestedPropertyToIntegerCast" and StringToIntegerCastFails failed with `Convert.ChangeType`
Such as
Convert.ChangeType(9.89m, typeof(int)) = 10, (int)9.89m = 9
Convert.ChangeType(true, typeof(int)) = 1, (int)true = fails

I also tried with T Cast<T>(object value) => return (T)value; store static instance of CastMethodInfo and make generic method before invoke but this was also failed. Which may run for primitive cases, for performance i am not sure of it? @hazzik